src/main/java/com/scalar/database/storage/cassandra4driver/StatementHandler.java

```diff
@@ -0,0 +1,146 @@
+package com.scalar.database.storage.cassandra4driver;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.datastax.oss.driver.api.core.ConsistencyLevel;
+import com.datastax.oss.driver.api.core.cql.BoundStatement;
+import com.datastax.oss.driver.api.core.cql.BoundStatementBuilder;
+import com.datastax.oss.driver.api.core.cql.PreparedStatement;
+import com.datastax.oss.driver.api.core.cql.ResultSet;
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.google.common.base.Joiner;
+import com.scalar.database.api.Consistency;
+import com.scalar.database.api.Operation;
+import com.scalar.database.api.Selection;
+import com.scalar.database.exception.storage.ExecutionException;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A handler class for statements */
+@ThreadSafe
+public abstract class StatementHandler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(StatementHandler.class);
+  protected final CqlSession session;
+
+  /**
+   * Constructs a {@code StatementHandler} with the specified {@link CqlSession}
+   *
+   * @param session {@code CqlSession}
+   */
+  protected StatementHandler(CqlSession session) {
+    this.session = checkNotNull(session);
+  }
+
+  /**
+   * Executes the specified {@code Operation}
+   *
+   * @param operation an {@code Operation} to execute
+   * @return a {@code ResultSet}
+   * @throws ExecutionException if the execution failed
+   */
+  @Nonnull
+  public ResultSet handle(Operation operation) throws ExecutionException {
+    try {
+      LOGGER.debug(operation + " is started.");
+      ResultSet results = handleInternal(operation);
+      LOGGER.debug(operation + " is executed normally.");
+      return results;
+
+    } catch (RuntimeException e) {
+      LOGGER.error(e.getMessage());
+      throw new ExecutionException(e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Executes the specified {@code Operation}
+   *
+   * @param operation an {@code Operation} to execute
+   * @return a {@code ResultSet}
+   */
+  @Nonnull
+  protected ResultSet handleInternal(Operation operation) {
+    PreparedStatement prepared = prepare(operation);
+    BoundStatementBuilder builder = bind(prepared, operation);
+    setConsistency(builder, operation);
+    return execute(builder.build(), operation);
+  }
+
+  /**
+   * Returns a {@link PreparedStatement} based on the given query string
+   *
+   * @param queryString
+   * @return a {@code PreparedStatement}
+   */
+  @Nonnull
+  protected PreparedStatement prepare(String queryString) {
+    LOGGER.debug("query to prepare : [" + queryString + "].");
+    return session.prepare(queryString);
```

Comment:
I'm just curious. Is there any specific reason why it does't cache prepared statements as before ?
It used to warn if it prepares the same query multiple times. It doesn't warn any more ?
The doc seems still saying this.
> Preparing the same query string twice is suboptimal and a bad practice, and will cause the driver to log a warning.

https://docs.datastax.com/en/drivers/java/4.0/com/datastax/oss/driver/api/core/cql/PreparedStatement.html

Reply:
Thank you!
According to the upgrade guide, prepared statements are cached.
I'll check the implementation.
https://github.com/datastax/java-driver/tree/4.x/upgrade_guide#statements

Reply:
Thanks ! That's cool ! It should have been done in the library.

Reply:
I've confirmed that the preparation process checks the cache before making a new prepared statement.

https://github.com/datastax/java-driver/blob/46817dfa60a70b0c22f926600efe39cbe7891435/core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlPrepareAsyncProcessor.java#L54-L83

Reply:
Cool ! Thanks.